compiler/src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -149,34 +149,42 @@ class TreeChecker extends Phase with SymTransformer {
 
     def withDefinedSym[T](tree: untpd.Tree)(op: => T)(implicit ctx: Context): T = tree match {
       case tree: untpd.DefTree =>
-        val sym = tree.symbol
```

Comment:
I think the original logic was complicated - that foldRightBN is sneaky because it mixes laziness with side-effects ! The changes remove the foldRightBN complexity but make everything else more complex. Proposal: make `withDefinedSyms` take a repeated argument and remove withDefinedSym. I.e.

       def withDefinedSyms[T](trees: untpd.Tree*)(op: => T)(implicit ctx: Context)

`withDefinedSyms` can step iteratively through all the trees, record the symbols found as defined in a ListBuffer or List and then execute `op`.